Fix 50227 malformed plists

sheagcraig opened the pull request:
### What does this PR do?
This PR updates the tests for the mac_utils module to not incorrectly pass malformed binary plists and then fixes the handling of that situation in the mac_utils module itself.

### What issues does this PR fix or reference?
#50227 

### Previous Behavior
Raised unhandled `plistlib.InvalidFileException`.

### New Behavior
Skips the poorly formed plist file by `continue`-ing the for loop.

### Tests written?

Yes (updated)

### Commits signed with GPG?

No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

sheagcraig committed:
Make test correctly generate a double exception.

sheagcraig committed:
Add secondary exception handler for malformed plists.

cachedout commented on salt/utils/mac_utils.py:
```diff
@@ -358,15 +358,21 @@ def _available_services(refresh=False):
+                        else:
+                            plist = plistlib.loads(
+                                salt.utils.stringutils.to_bytes(plist_xml))
+                    except Exception:
```
We usually don't like bare excepts. Do you know the specific type(s) of exceptions that can be expected here?

sheagcraig replied:
Hah, I'm with you on that. I actually put that in there because the previous block used `Exception`. I thought that was a little strange. I'll push some changes to specify the actually expected exceptions tomorrow.

sheagcraig commented:
@cachedout and anybody else, this is a little more involved than I thought, and I need some guidance.

Two main questions:
1. It appears to me like the `available_services` function as written is adding in a service even if the plist file is malformed. I doubt in practice that you could ever load a service if the plist file was invalid XML, so I'm not sure this is useful or expected behavior. My instinct would be to not show a service in the list if it's not possible to load it (due to invalid XML).
2. The tests seem to affirm this, and they seem to be copy & paste duplicates from the other tests, with just minor edits for each test. But with more explicit exception handling, I've got at least 3 different "malformed XML" situations to test for (py2, py3 completely invalid, py3 broken tags / mostly valid XML). Is the goal to keep the tests essentially the same, or do I have the ability to break them up into more tests?

sheagcraig commented:
Hi-wondering if I can get some clarification on the following point. Currently the mac_utils services list includes services which can't actually start due to invalid plist files. The code handles this by falling back to just using the filename to populate the list of services (rather than the ID property from inside the file).

This seems wrong to me, and I'd be fine with updating the code to exclude services which can't start from the list of available services. This involves rewriting some tests as well.

OR, if the goal is still to include those services, I can revert back mac_utils and just fix the exception handling for malformed XML cases and update the tests to pass with the existing behavior.

rallytime commented:
@Ch3LL or @weswhet can you help out with the questions above?

weswhet commented:
@sheagcraig I would say go for if the plist is malformed to NOT include it in the available services. As it technically wouldn't be available. I would probably throw a warning or error to let folks know that it's a bad file. 

@rallytime could we add this into 2018 as it is a bug fix?

rallytime commented:
@weswhet Yes, I've marked it for back-porting. Thanks!

sheagcraig commented:
I'll get on finishing this up today then. Thanks for the feedback!

sheagcraig committed:
Avoid `AttributeError` when checking for exception types.

sheagcraig committed:
Handle bad plists by ignoring them, with py2 and py3 specific approaches

sheagcraig committed:
Fix trailing-slash bug in User LaunchAgent appending.

I discovered this issue while updating the tests, so there's no open
issue for it.

sheagcraig committed:
Update testing for mac_utils services.

This commit begins to go over the existing testing for mac_utils,
applying the updated best-practices from the unit testing section of the
Salt docs.

Wherever possible, I've tried to keep the data for each test within the
test, while factoring out repetitive functions. In this commit, for
example, the `os.walk` side_effect func has been factored into its own
helper function.

This commit begins looking at what is being tested for this module and
making it more specific, specifically to test one thing per test
(existing tests had multiple asserts and in some casaes were actually
malformed such that the tests would pass, but in actual use, code would
throw an exception and skip over the block).

sheagcraig committed:
Factor out another helper function, implement broken symlink test.

sheagcraig committed:
Condense service result assertions, add a binary plist test.

sheagcraig committed:
Write tests to handle invalid launchd plist XML.

sheagcraig committed:
Fix linting errors.

sheagcraig commented:
Okay-I think this is ready for a look again.

weswhet commented:
LGTM!!! Thanks, bud!

cachedout reviewed (approved):
LGTM. Nice work here, @sheagcraig
